File: AlgoToolbox/week2/fibo_huge.cpp
#include<iostream>

using namespace std;

int period(int m){
	int r = 1;
	int a =0, b = 1 , c;
	do{
		c = (a+b)%m;
		a = b;
		b = c;
		r++;
	}while(!(c==0 && a==1));
	return r;
}

uint64_t fibo_last(uint64_t n){
	if(n <= 1)
		return n;
	else{
		uint64_t a=0UL , b=1UL , c;
		uint64_t i =2UL;
		while(i <= n){
			c = a + b;
			a = b;
			b = c;
			i = i + 1UL;
		}
		return c;
	}
}

int fibo_mod(uint64_t n, int m){
	int p = period(m);
	return fibo_last(n%p)%m;
}

int main(){
	uint64_t n;
	int m;
	cin>>n>>m;
	cout<<fibo_mod(n , m)<<endl;
	return 0;
}

File: AlgoToolbox/week2/fibo_huge.py
#python3
def period(m):
    r = 1
    a = 0
    b = 1
    c = 0
    while(not (c == 0 and a == 1)):
        c = (a+b)%m
        a = b
        b = c
        r = r + 1
    return r

def fibo(n):
    if n <= 1:
        return n
    else:
        a = 0
        b = 1
        c = 0
        for i in range(2, n + 1):
            c = a + b
            a = b
            b = c
        return c

def fibo_mod(n, m):
    p = period(m)
    return fibo(n%p)%m

a = list()
a = input().split()

n = int(a[0])
m = int(a[1])

r = fibo_mod(n,m)
print(r)


